dockerfile first draft

choldgraf opened the pull request:
hey @yuvipanda , here's a first take at converting your notes to a proper doc page. It's still WIP but I'd appreciate your thoughts on it thus far. There are a few places in the notes where I need to fill in some info, and I'm also still not positive that this repository is the right place for this, but we can discuss :-)

http://predictablynoisy.com/binder/

TODO:

- [x] Add a section on reproducibility with Dockerfiles

minrk commented:
It would be great to add at the top, something pointing to "If you find you have to use a Dockerfile, please let us know why. We hope to get to a point where nobody ever has to resort to a custom Dockerfile."

choldgraf commented:
@minrk can you write down some quick points to cover re: reproducibility with dockerfiles? I'll flesh it out and add to this PR

minrk commented:
Points re: reproducibility with Dockerfiles:

1. Always tag your base image! Use `FROM base/image:tag` instead of `FROM base/image`. The default (`:latest`) tag is ~guaranteed to change in breaking ways as upstream software gets updates. What tag to use depends on the tagging scheme of your base image. Docker tags are like git branches and can change. The safest choice is to use docker sha tags, e.g.

       FROM ubuntu@sha256:84c334414...

   which are more like git commits in that they are guaranteed to refer to a single version of the image.

2. Binder's requirements for Dockerfiles are in beta and changing, so Dockerfiles on Binder **will break** from time to time until Binder solidifies its requirements. Dockerfiles are not an appropriate choice for reproducibility with Binder at the moment.

And re: reproducibility in general:

1. Pin your dependencies when installing. When you install a dependency, don't just specify `numpy`, specify `numpy==1.12.0`. `pip freeze` is a handy tool to export the exact version of every Python package in your environment in a format that can be used in `requirements.txt`. `conda env export -n <env-name>` is the equivalent for conda `environment.yml`.

choldgraf commented:
ok, added the language that @minrk suggested and cleaned things up a bit...think we can merge this? It definitely still has WIP components to it, but I think it does have valuable info

minrk commented:
👍 @willingc do you want to give it a pass and then merge?

willingc commented on doc/dockerfile.rst:
```diff
@@ -0,0 +1,139 @@
+When you are building complex software
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+Some software is simply too complex to be specified with Binder configuration
```
Most Binder configurations can be achieved without a Dockerfile. Before resorting to a Dockerfile, we recommend trying to use ``postBuild`` commands for configuration.

willingc commented on doc/dockerfile.rst:
```diff
@@ -0,0 +1,139 @@
+a different language, it may be possible to accomplish this with a Dockerfile.
+
+For a list of languages that Binder supports with configuration files, see
+``the repo2docker documentation <http://repo2docker.readthedocs.io/en/latest/>`_.
```
` Extra backtick.

willingc commented on doc/dockerfile.rst:
```diff
@@ -0,0 +1,139 @@
+Preparing your Dockerfile
+=========================
+
+For a Dockerfile to work on binder, it must meet the following requirements:
```
Binder

willingc commented on doc/dockerfile.rst:
```diff
@@ -0,0 +1,139 @@
+   naming conventions differ between images, so we recommend using
+   the SHA tag of the image.
+
+   Here's an example of a Dockerfile ``FROM`` statement that would work:
```
... work.

choldgraf commented:
ok @willingc updates made, LMK

willingc commented on doc/index.rst:
```diff
@@ -95,3 +95,20 @@ package in your environment in a format that can be used in ``requirements.txt``
 ``conda env export -n <env-name>`` is the equivalent for anaconda's environment.yml
 file.
+
+If using a Dockerfile, source a reproducible, tagged image
```
Let's move all of this over to the `dockerfile.rst` document. My understanding is that we are trying to guide folks to using Dockerfiles as a choice in a very limited use case (i.e. when pip and conda can't get you what is needed). I think it's fine to have a short paragraph or a few bullets here, but the current text is mixing implementation with general tips for reproducibility.

choldgraf replied:
+1

choldgraf commented:
updated per @willingc 's comments...LMK if you'd like me to change some other language in there. I think we should probably have another PR to flesh out the reproducibility tips more in the future (maybe when we create the reproducibility validator machinery)

willingc commented on doc/index.rst:
``````````````````diff
@@ -96,19 +96,9 @@ package in your environment in a format that can be used in ``requirements.txt``
+Using Dockerfiles
+`````````````````
 
+Ensuring reproducibility with Dockerfiles comes with its own set of challenges.
``````````````````
Love this wording.

willingc commented:
Thanks @choldgraf for the edits. It looks good.

choldgraf commented:
Yay!
